torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2812,13 +2821,440 @@ Tensor _cudnn_ctc_loss_backward(const Tensor& grad_out, const Tensor& loss, cons
   }
 }
 
-bool any_variable_defined(variable_list& variables) {
-  for (auto variable : variables) {
-    if (variable.defined()) {
-      return true;
+Tensor mkldnn_convolution_forward(const Tensor& self_fw_grad, const Tensor& weight_fw_grad, const Tensor& bias_fw_grad,
+        const Tensor& self, const Tensor& weight, IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int groups,
+        const Tensor& result) {
+  Tensor out_fw_grad;
+
+  if (self_fw_grad.defined()) {
+    out_fw_grad = at::mkldnn_convolution(self_fw_grad, weight, Tensor(), padding, stride, dilation, groups);
+  }
+
+  if (weight_fw_grad.defined()) {
+    auto val = at::mkldnn_convolution(self, weight_fw_grad, bias_fw_grad, padding, stride, dilation, groups);
+    if (out_fw_grad.defined()) {
+      out_fw_grad = out_fw_grad + val;
+    } else {
+      out_fw_grad = val;
+    }
+  } else if (bias_fw_grad.defined()) {
+    // Used to make the broadcasting work when working with channel-only values
+    auto view_size = result.sizes().vec();
+    for (auto dim = 0; dim < view_size.size(); ++dim) {
+      // Don't change the channel size
+      if (dim != 1) {
+        view_size[dim] = 1;
+      }
+    }
+
+    if (out_fw_grad.defined()) {
+      out_fw_grad = out_fw_grad + bias_fw_grad.view(view_size);
+    } else {
+      out_fw_grad = bias_fw_grad.view(view_size).expand_as(result);
     }
   }
-  return false;
+
+  return out_fw_grad;
+}
+
+Tensor thnn_conv2d_forward_grad(const Tensor& self_fw_grad, const Tensor& weight_fw_grad, IntArrayRef kernel_size, const Tensor& bias_fw_grad,
+        const Tensor& self, const Tensor& weight, IntArrayRef stride, IntArrayRef padding, const Tensor& result) {
+  Tensor out_fw_grad;
+
+  if (self_fw_grad.defined()) {
+    out_fw_grad = std::get<0>(at::thnn_conv2d_forward(self_fw_grad, weight, kernel_size, Tensor(), stride, padding));
+  }
+
+  if (weight_fw_grad.defined()) {
+    auto val = std::get<0>(at::thnn_conv2d_forward(self, weight_fw_grad, kernel_size, bias_fw_grad, stride, padding));
+    if (out_fw_grad.defined()) {
+      out_fw_grad = out_fw_grad + val;
+    } else {
+      out_fw_grad = val;
+    }
+  } else if (bias_fw_grad.defined()) {
+    // Used to make the broadcasting work when working with channel-only values
+    auto view_size = result.sizes().vec();
+    for (auto dim = 0; dim < view_size.size(); ++dim) {
+      // Don't change the channel size
+      if (dim != 1) {
+        view_size[dim] = 1;
+      }
+    }
+
+    if (out_fw_grad.defined()) {
+      out_fw_grad = out_fw_grad + bias_fw_grad.view(view_size);
+    } else {
+      out_fw_grad = bias_fw_grad.view(view_size).expand_as(result);
+    }
+  }
+
+  return out_fw_grad;
+}
+
+Tensor slow_conv_dilated2d_forward_grad(const Tensor& self_fw_grad, const Tensor& weight_fw_grad, IntArrayRef kernel_size, const Tensor& bias_fw_grad,
```

Comment:
The copy pasting here seems like a big problem. What's the plan?

Reply:
The end plan is to refactor these functions and make them easier to read using a macro.
I leave that for a follow up as I prefer to have more formulas before deciding on a final strategy here.